### PR TITLE
Crazy price 005 validators interfaces logger

### PR DESCRIPTION
**Pull request by MACE4GITHUB**
Updated Interfaces, Validators, Tests. 
Added Language support.
Added Logger.

### CODE REVIEW

File: src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Configurations/ICrazyPriceLogger.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+﻿using Microsoft.Extensions.Hosting;
+
+namespace Exadel.CrazyPrice.Common.Configurations
```
Comment (yegorl) [inline src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Configurations/ICrazyPriceLogger.cs:3]: Are you sure that we need this file? In a case of 2 loggers no sense to create extra interface

File: src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/CrazyPriceHost.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+            if (!hasLogging)
+            {
+                hostBuilder.Build().Run();
+                return;
+            }
+
+            var usedLogger = configuration.GetSection(loggerName).Value.ToLowerInvariant();
+
+            ICrazyPriceLogger logger = usedLogger switch
```
Comment (yegorl) [inline src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/CrazyPriceHost.cs:38]: Rework to static instances - no need for interfaces here

File: src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Configurations/NLogLogger.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+        public NLogLogger(IHostBuilder hostBuilder, IConfigurationRoot configuration)
+        {
+            _hostBuilder = hostBuilder;
+            _configuration = configuration;
+        }
+
+        public IHostBuilder UseLogger()
+        {
+            LogManager.Configuration = new NLogLoggingConfiguration(_configuration.GetSection("NLog"));
```
Comment (yegorl) [inline src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Configurations/NLogLogger.cs:26]: Configuration is specific for each logger - remove this logic to upper level

File: src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/CrazyPriceHost.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+using System.IO;
+
+namespace Exadel.CrazyPrice.Common
+{
+    public static class CrazyPriceHost
+    {
+        public static void CheckLoggerAndStartHost(IHostBuilder hostBuilder,
+            string initMessage, string errorInitMessage)
+        {
+            const string useLogging = "Logging:UseLogging";
```
Comment (yegorl) [inline src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/CrazyPriceHost.cs:14]: No sense to create constants here - put them to class or just put in code.

File: src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/CrazyPriceHost.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+            const string nLog = "nlog";
+            const string serilog = "serilog";
+            const string appsettings = "appsettings.json";
+
+            var configuration = new ConfigurationBuilder()
+                .SetBasePath(Directory.GetCurrentDirectory())
+                .AddJsonFile(appsettings)
+                .Build();
+
+            if (!bool.TryParse(configuration.GetSection(useLogging).Value, out var hasLogging))
```
Comment (yegorl) [inline src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/CrazyPriceHost.cs:25]: Create methods for reading configuration parameters which will raise exception if no required data found. Also create class for Serilog/Nlog configuration and fill them here and pass to each logger own class

File: src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/CrazyPriceHost.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+using Microsoft.Extensions.Hosting;
+using System;
+using System.IO;
+
+namespace Exadel.CrazyPrice.Common
+{
+    public static class CrazyPriceHost
+    {
+        public static void CheckLoggerAndStartHost(IHostBuilder hostBuilder,
+            string initMessage, string errorInitMessage)
```
Comment (yegorl) [inline src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/CrazyPriceHost.cs:12]: Why you need init<essage/initError messages

It makes sense to run Logger only here - if something get wrong exception will raised and app will stop. 
Run in separate place. Rename method as well

**Review by yegorl — CHANGES_REQUESTED** [review]

File: src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Configurations/ICrazyPriceLogger.cs (hunk 1 above)
Comment (MACE4GITHUB) [reply src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Configurations/ICrazyPriceLogger.cs]: Removed.

File: src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/CrazyPriceHost.cs (hunk 2 above)
Comment (MACE4GITHUB) [reply src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/CrazyPriceHost.cs]: Removed.

File: src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Configurations/NLogLogger.cs (hunk 3 above)
Comment (MACE4GITHUB) [reply src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Configurations/NLogLogger.cs]: Removed.

File: src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/CrazyPriceHost.cs (hunk 5 above)
Comment (MACE4GITHUB) [reply src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/CrazyPriceHost.cs]: Removed.

File: src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/CrazyPriceHost.cs (hunk 6 above)
Comment (MACE4GITHUB) [reply src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/CrazyPriceHost.cs]: Removed.

File: src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/CrazyPriceHost.cs (hunk 4 above)
Comment (MACE4GITHUB) [reply src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/CrazyPriceHost.cs]: Removed.

File: src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Extentions/HostBuilderExtentions.cs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+                if (HasLog(hostingContext, "LogToNLog"))
+                {
+                    logging.AddNLog();
+                }
+            });
+        }
+
+        private static bool HasLog(HostBuilderContext hostingContext, string kindLod)
+        {
+            if (!bool.TryParse(hostingContext.Configuration.GetSection(kindLod).Value, out var hasLog))
```
Comment (yegorl) [inline src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Extentions/HostBuilderExtentions.cs:35]: Exception can be raised here if no such section - please check

File: src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Extentions/HostBuilderExtentions.cs [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+using Microsoft.Extensions.Logging;
+using NLog.Extensions.Logging;
+using Serilog;
+using System;
+
+namespace Exadel.CrazyPrice.Common.Extentions
+{
+    public static class HostBuilderExtentions
+    {
+        public static IHostBuilder UseLogger(this IHostBuilder hostBuilder)
```
Comment (yegorl) [inline src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Extentions/HostBuilderExtentions.cs:11]: Probably rename to SetupLogger

File: src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Extentions/EnumExtentions.cs [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    public static class EnumExtentions
+    {
+        /// <summary>
+        /// Gets the formatted string from enum.
+        /// $k or $K - the key enum,
+        /// $v or $V - the value enum.
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="formatString"></param>
+        /// <returns></returns>
```
Comment (yegorl) [inline src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Extentions/EnumExtentions.cs:18]: Check - seems ASP.NET will do this automatically - no need to worry. Just check in validator InEnum

**Review by yegorl — CHANGES_REQUESTED** [review]

File: src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Extentions/HostBuilderExtentions.cs (hunk 8 above)
Comment (MACE4GITHUB) [reply src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Extentions/HostBuilderExtentions.cs]: [Done](https://github.com/yegorl/exadel_discounts_be/pull/6/commits/d20afd5e924712fe206173f9b892ca771e6e6adf)

File: src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Extentions/EnumExtentions.cs (hunk 9 above)
Comment (MACE4GITHUB) [reply src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Extentions/EnumExtentions.cs]: [Done](https://github.com/yegorl/exadel_discounts_be/pull/6/commits/92b05436440d8d29c426e3b8d7b0f62ea50d0784)

File: src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Extentions/HostBuilderExtentions.cs (hunk 7 above)
Comment (MACE4GITHUB) [reply src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Extentions/HostBuilderExtentions.cs:35]: It should be.

File: src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Extentions/HostBuilderExtentions.cs [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+                {
+                    Environment.SetEnvironmentVariable("BASEDIR", AppDomain.CurrentDomain.BaseDirectory);
+                    logging.AddSerilog(new LoggerConfiguration()
+                        .ReadFrom.Configuration(hostingContext.Configuration)
+                        .CreateLogger());
+                }
+
+                if (HasLog(hostingContext, "LogToNLog"))
+                {
+                    logging.AddNLog();
```
Comment (yegorl) [inline src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Extentions/HostBuilderExtentions.cs:28]: Does NLog standart parameters with nlog.config used here?

File: src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Extentions/StringExtentions.cs [hunk 11]
```diff
@@ -10,77 +12,207 @@ namespace Exadel.CrazyPrice.Common.Extentions
...
 
         /// <summary>
-        /// Removes 2 and more space. Trim.
+        /// Gets string what two and more chars replaced by some one.
         /// </summary>
         /// <param name="s"></param>
+        /// <param name="chars"></param>
         /// <returns></returns>
-        public static string ReplaceTwoAndMoreSpaceByOne(this string s)
+        public static string ReplaceTwoAndMoreCharsBySomeOne(this string s, char[] chars)
```
Comment (yegorl) [inline src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Extentions/StringExtentions.cs:68]: Probably string to replace? Do we need this method?

File: src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Extentions/StringExtentions.cs [hunk 12]
```diff
@@ -10,77 +12,207 @@ namespace Exadel.CrazyPrice.Common.Extentions
...
+            if (chars == null || chars.Length == 0)
             {
-                return s;
+                throw new ArgumentNullException(nameof(chars));
             }
 
-            s = ' ' + s;
-            var stbBuilder = new StringBuilder();
-            for (var i = 1; i <= s.Length - 1; i++)
+            s = s.Trim(chars);
```
Comment (yegorl) [inline src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Extentions/StringExtentions.cs:80]: If I need to replace with space?

**Review by yegorl — APPROVED** [review]

File: src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Extentions/HostBuilderExtentions.cs (hunk 10 above)
Comment (MACE4GITHUB) [reply src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Extentions/HostBuilderExtentions.cs]: logging.AddNLog() changed to logging.AddNLogWeb() for support template like ${aspnet-request-method}

File: src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Extentions/StringExtentions.cs (hunk 12 above)
Comment (MACE4GITHUB) [reply src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Extentions/StringExtentions.cs:80]: It replaces two and more chars by some one. If spaces by space ("      " to " "), if "---" by "-". It works wery fast.

**Review by yegorl — APPROVED** [review]

File: src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Extentions/StringExtentions.cs (hunk 11 above)
Comment (MACE4GITHUB) [reply src/Exadel.CrazyPrice/Exadel.CrazyPrice.Common/Extentions/StringExtentions.cs:68]: Used by GetValidContent.

**Pull request merged**